File: tools/calendar/gcal_quick.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import {authenticate} from '@google-cloud/local-auth';
import {google} from 'googleapis';

const SCOPES = ['https://www.googleapis.com/auth/calendar'];
const BASE = path.resolve(process.cwd(), 'tools/calendar');
const TOKEN_PATH = path.join(BASE, 'token.json');
const CREDENTIALS_PATH = path.join(BASE, 'credentials.json');

async function loadSavedCredentialsIfExist() {
  if (!fs.existsSync(TOKEN_PATH)) return null;
  const content = fs.readFileSync(TOKEN_PATH, 'utf8');
  const credentials = JSON.parse(content);
  return google.auth.fromJSON(credentials);
}

function saveCredentials(client) {
  const keys = JSON.parse(fs.readFileSync(CREDENTIALS_PATH, 'utf8'));
  const key = keys.installed || keys.web;
  const payload = JSON.stringify({
    type: 'authorized_user',
    client_id: key.client_id,
    client_secret: key.client_secret,
    refresh_token: client.credentials.refresh_token,
  });
  fs.writeFileSync(TOKEN_PATH, payload);
}

async function authorize() {
  let client = await loadSavedCredentialsIfExist();
  if (!client) {
    if (!fs.existsSync(CREDENTIALS_PATH)) {
      throw new Error(`Missing OAuth credentials: ${CREDENTIALS_PATH}`);
    }
    client = await authenticate({
      scopes: SCOPES,
      keyfilePath: CREDENTIALS_PATH,
    });
    if (client.credentials) saveCredentials(client);
  }
  return client;
}

function arg(name, fallback = null) {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 ? process.argv[i + 1] : fallback;
}

async function main() {
  const cmd = process.argv[2];
  const auth = await authorize();
  const calendar = google.calendar({version: 'v3', auth});
  const calendarId = arg('calendar', 'primary');

  if (cmd === 'create') {
    const title = arg('title');
    const description = arg('description', '');
    const timezone = arg('timezone', 'Australia/Adelaide');
    const allDay = process.argv.includes('--all-day');
    const body = {summary: title, description};
    if (allDay) {
      body.start = {date: arg('start-date')};
      body.end = {date: arg('end-date')};
    } else {
      body.start = {dateTime: arg('start'), timeZone: timezone};
      body.end = {dateTime: arg('end'), timeZone: timezone};
    }
    const res = await calendar.events.insert({calendarId, requestBody: body});
    console.log(JSON.stringify({id: res.data.id, htmlLink: res.data.htmlLink}));
    return;
  }

  if (cmd === 'update') {
    const eventId = arg('event-id');
    const timezone = arg('timezone', 'Australia/Adelaide');
    const getRes = await calendar.events.get({calendarId, eventId});
    const event = getRes.data;
    const title = arg('title');
    if (title) event.summary = title;
    const description = arg('description');
    if (description !== null) event.description = description;
    const start = arg('start');
    const end = arg('end');
    if (start && end) {
      event.start = {dateTime: start, timeZone: timezone};
      event.end = {dateTime: end, timeZone: timezone};
    }
    const res = await calendar.events.update({calendarId, eventId, requestBody: event});
    console.log(JSON.stringify({id: res.data.id, htmlLink: res.data.htmlLink}));
    return;
  }

  if (cmd === 'list') {
    const timeMin = arg('time-min');
    const timeMax = arg('time-max');
    const maxResults = Number(arg('limit', '20'));
    const res = await calendar.events.list({
      calendarId,
      timeMin,
      timeMax,
      singleEvents: true,
      orderBy: 'startTime',
      maxResults,
    });
    for (const e of res.data.items || []) {
      const start = e.start?.dateTime || e.start?.date;
      console.log(`${e.id}\t${start}\t${e.summary || '(no title)'}`);
    }
    return;
  }

  throw new Error('Usage: gcal_quick.mjs <create|update|list> ...');
}

main().catch((e) => {
  console.error(e.message || e);
  process.exit(1);
});

File: tools/calendar/gcal_quick.py
#!/usr/bin/env python3
import argparse
import datetime as dt
import json
import os
from pathlib import Path

from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build

SCOPES = ["https://www.googleapis.com/auth/calendar"]
BASE = Path.home() / ".openclaw" / "workspace" / "tools" / "calendar"
TOKEN_PATH = BASE / "token.json"
CREDS_PATH = BASE / "credentials.json"


def get_service():
    creds = None
    if TOKEN_PATH.exists():
        creds = Credentials.from_authorized_user_file(str(TOKEN_PATH), SCOPES)
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            if not CREDS_PATH.exists():
                raise SystemExit(
                    f"Missing credentials file: {CREDS_PATH}\n"
                    "Download OAuth client JSON from Google Cloud Console and place it there."
                )
            flow = InstalledAppFlow.from_client_secrets_file(str(CREDS_PATH), SCOPES)
            creds = flow.run_local_server(port=0)
        TOKEN_PATH.write_text(creds.to_json())
    return build("calendar", "v3", credentials=creds)


def parse_iso(s: str) -> str:
    # Expect: 2026-03-22T17:00:00+10:30
    return dt.datetime.fromisoformat(s).isoformat()


def cmd_create(args):
    service = get_service()
    event = {
        "summary": args.title,
        "description": args.description or "",
    }
    if args.all_day:
        event["start"] = {"date": args.start_date}
        event["end"] = {"date": args.end_date}
    else:
        event["start"] = {"dateTime": parse_iso(args.start), "timeZone": args.timezone}
        event["end"] = {"dateTime": parse_iso(args.end), "timeZone": args.timezone}

    created = service.events().insert(calendarId=args.calendar, body=event).execute()
    print(json.dumps({"id": created["id"], "htmlLink": created.get("htmlLink")}, ensure_ascii=False))


def cmd_update(args):
    service = get_service()
    event = service.events().get(calendarId=args.calendar, eventId=args.event_id).execute()
    if args.title:
        event["summary"] = args.title
    if args.description is not None:
        event["description"] = args.description
    if args.start and args.end:
        event["start"] = {"dateTime": parse_iso(args.start), "timeZone": args.timezone}
        event["end"] = {"dateTime": parse_iso(args.end), "timeZone": args.timezone}
    updated = service.events().update(calendarId=args.calendar, eventId=args.event_id, body=event).execute()
    print(json.dumps({"id": updated["id"], "htmlLink": updated.get("htmlLink")}, ensure_ascii=False))


def cmd_list(args):
    service = get_service()
    events = (
        service.events()
        .list(
            calendarId=args.calendar,
            timeMin=args.time_min,
            timeMax=args.time_max,
            singleEvents=True,
            orderBy="startTime",
            maxResults=args.limit,
        )
        .execute()
        .get("items", [])
    )
    for e in events:
        start = e.get("start", {}).get("dateTime") or e.get("start", {}).get("date")
        print(f"{e.get('id')}\t{start}\t{e.get('summary','(no title)')}")


def main():
    p = argparse.ArgumentParser()
    sub = p.add_subparsers(required=True)

    c = sub.add_parser("create")
    c.add_argument("--calendar", default="primary")
    c.add_argument("--title", required=True)
    c.add_argument("--description")
    c.add_argument("--timezone", default="Australia/Adelaide")
    c.add_argument("--start")
    c.add_argument("--end")
    c.add_argument("--all-day", action="store_true")
    c.add_argument("--start-date")
    c.add_argument("--end-date")
    c.set_defaults(func=cmd_create)

    u = sub.add_parser("update")
    u.add_argument("--calendar", default="primary")
    u.add_argument("--event-id", required=True)
    u.add_argument("--title")
    u.add_argument("--description")
    u.add_argument("--timezone", default="Australia/Adelaide")
    u.add_argument("--start")
    u.add_argument("--end")
    u.set_defaults(func=cmd_update)

    l = sub.add_parser("list")
    l.add_argument("--calendar", default="primary")
    l.add_argument("--time-min", required=True)
    l.add_argument("--time-max", required=True)
    l.add_argument("--limit", type=int, default=20)
    l.set_defaults(func=cmd_list)

    args = p.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()
